source/sundials/sunlinsol_wrapper.cc

```diff
@@ -0,0 +1,323 @@
+//
+//    The deal.II library is free software; you can use it, redistribute
+//    it, and/or modify it under the terms of the GNU Lesser General
+//    Public License as published by the Free Software Foundation; either
+//    version 2.1 of the License, or (at your option) any later version.
+//    The full text of the license can be found in the file LICENSE.md at
+//    the top level directory of deal.II.
+//
+//-----------------------------------------------------------
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/sundials/sunlinsol_wrapper.h>
+
+#ifdef DEAL_II_WITH_SUNDIALS
+#  if DEAL_II_SUNDIALS_VERSION_GTE(4, 0, 0)
+
+#    include <deal.II/base/exceptions.h>
+
+#    include <deal.II/lac/block_vector.h>
+#    include <deal.II/lac/la_parallel_block_vector.h>
+#    include <deal.II/lac/la_parallel_vector.h>
+#    include <deal.II/lac/vector.h>
+#    ifdef DEAL_II_WITH_TRILINOS
+#      include <deal.II/lac/trilinos_parallel_block_vector.h>
+#      include <deal.II/lac/trilinos_vector.h>
+#    endif
+#    ifdef DEAL_II_WITH_PETSC
+#      include <deal.II/lac/petsc_block_vector.h>
+#      include <deal.II/lac/petsc_vector.h>
+#    endif
+
+#    include <deal.II/sundials/n_vector.h>
+
+#    if DEAL_II_SUNDIALS_VERSION_LT(5, 0, 0)
+#      include <deal.II/sundials/sunlinsol_newempty.h>
+#    endif
+
+DEAL_II_NAMESPACE_OPEN
+
+
+
+namespace SUNDIALS
+{
+  DeclException1(ExcSundialsSolverError,
+                 int,
+                 << "One of the SUNDIALS linear solver internal"
+                 << " functions returned a negative error code: " << arg1
+                 << ". Please consult SUNDIALS manual.");
+
+#    define AssertSundialsSolver(code) \
+      Assert(code >= 0, ExcSundialsSolverError(code))
+
+  namespace internal
+  {
+    /**
+     * storage for internal content of the linear solver wrapper
+     */
+    template <typename VectorType>
+    struct LinearSolverContent
+    {
+      ATimesFn a_times_fn;
+      PSetupFn preconditioner_setup;
+      PSolveFn preconditioner_solve;
+
+      LinearSolveFunction<VectorType> lsolve;
+
+      void *P_data;
+      void *A_data;
+    };
+  } // namespace internal
+
+  namespace
+  {
+    using namespace SUNDIALS::internal;
+
+    /**
+     * Access our LinearSolverContent from the generic content of the
+     * SUNLinearSolver @p ls.
+     */
+    template <typename VectorType>
+    LinearSolverContent<VectorType> *
+    access_content(SUNLinearSolver ls)
+    {
+      Assert(ls->content != nullptr, ExcInternalError());
+      return static_cast<LinearSolverContent<VectorType> *>(ls->content);
+    }
+
+
+
+    SUNLinearSolver_Type arkode_linsol_get_type(SUNLinearSolver)
+    {
+      return SUNLINEARSOLVER_ITERATIVE;
+    }
+
+
+
+    template <typename VectorType>
+    int
+    arkode_linsol_solve(SUNLinearSolver LS,
+                        SUNMatrix,
+                        N_Vector x,
+                        N_Vector b,
+                        realtype tol)
```

Comment:
I have not checked the size of these objects and the cost it takes to copy them upon entering the function (I assume small because it is a C struct), but I still wonder if we should to pass them by reference rather than by value. I also saw pass-by-value in some other functions, so I guess that might not be too problematic in the end.

Reply:
Not sure if I understand what you mean. This specific function signature is mandated by ARKODE and we cannot change it. All the types (except realtype) are pointer types. Or which objects are you referring to?

Reply:
Right, this is what I wanted to ensure regarding the types. So we should be fine.